src/javaharness/java/arcs/webimpl/DeviceClientJsImpl.java

```diff
@@ -20,68 +21,60 @@
 
 public class DeviceClientJsImpl implements DeviceClient {
 
+    private final String FIELD_MESSAGE = "message";
+    private final String MESSAGE_READY = "ready";
+    private final String MESSAGE_DATA = "data";
+    private final String FIELD_TRANSACTION_ID = "tid";
+    private final String FIELD_DATA = "data";
+
     private final PortableJsonParser jsonParser;
-    private Map<String, ArcsEnvironment.SuggestionListener> inProgress;
+    private Map<String, ArcsEnvironment.DataListener> inProgress;
     private final PECInnerPortFactory portFactory;
     private final Map<String, PECInnerPort> portById = new HashMap<String, PECInnerPort>();
 
     @Inject
     public DeviceClientJsImpl(PortableJsonParser jsonParser,
-                              Map<String, ArcsEnvironment.SuggestionListener> inProgress,
+                              Map<String, ArcsEnvironment.DataListener> inProgress,
                               PECInnerPortFactory portFactory) {
         this.jsonParser = jsonParser;
         this.inProgress = inProgress;
         this.portFactory = portFactory;
     }
 
-    public void foundSuggestions(String transactionId, String content) {
-        try {
-//            window.alert("Found Suggestions2 called " + content);
-
-            LinkedHashSet<String> suggestions = new LinkedHashSet<>();
-
-            // We accept either an json array of objects or json object
-            if (content.startsWith("[")) {
-                PortableJson suggestionArray = jsonParser.parse(content);
-                if (suggestionArray.getLength() == 0) {
-                    return;
+    @Override
+    public void receive(String json) {
+        // window.console.log("receive called " + json);
+        PortableJson content = jsonParser.parse(json);
+        String message = content.getString(FIELD_MESSAGE);
+        switch (message) {
+            case MESSAGE_READY:
+                window.console.log("Received 'ready' message");
+                break;
+            case MESSAGE_DATA:
+                String transactionId = String.valueOf(content.getInt(FIELD_TRANSACTION_ID));
+                if (inProgress.containsKey(transactionId)) {
+                    inProgress.get(transactionId).onData(transactionId, jsonParser.stringify(content.getObject(FIELD_DATA)));
```

Comment:
I wonder if we need a null check for content.getObject() here?

Reply:
done